src/tds/codec/token/token_env_change.rs

```diff
@@ -151,82 +154,117 @@ impl TokenEnvChange {
 
         let token = match ty {
             EnvChangeTy::Database => {
-                let len = src.read_u8().await?;
-                let new_value = read_varchar(src, len).await?;
+                let len = buf.read_u8()? as usize;
+                let mut bytes = vec![0; len];
+
+                for i in 0..len {
+                    bytes[i] = buf.read_u16::<LittleEndian>()?;
+                }
 
-                let len = src.read_u8().await?;
-                let old_value = read_varchar(src, len).await?;
+                let new_value = String::from_utf16(&bytes[..])?;
+
+                let len = buf.read_u8()? as usize;
+                let mut bytes = vec![0; len];
+
+                for i in 0..len {
+                    bytes[i] = buf.read_u16::<LittleEndian>()?;
+                }
+
+                let old_value = String::from_utf16(&bytes[..])?;
 
                 TokenEnvChange::Database(new_value, old_value)
             }
             EnvChangeTy::PacketSize => {
-                let len = src.read_u8().await?;
-                let new_value = read_varchar(src, len).await?;
+                let len = buf.read_u8()? as usize;
+                let mut bytes = vec![0; len];
+
+                for i in 0..len {
+                    bytes[i] = buf.read_u16::<LittleEndian>()?;
+                }
+
+                let new_value = String::from_utf16(&bytes[..])?;
+
+                let len = buf.read_u8()? as usize;
+                let mut bytes = vec![0; len];
+
+                for i in 0..len {
+                    bytes[i] = buf.read_u16::<LittleEndian>()?;
+                }
 
-                let len = src.read_u8().await?;
-                let old_value = read_varchar(src, len).await?;
+                let old_value = String::from_utf16(&bytes[..])?;
 
                 TokenEnvChange::PacketSize(new_value.parse()?, old_value.parse()?)
             }
             EnvChangeTy::SqlCollation => {
-                let len = src.read_u8().await? as usize;
+                let len = buf.read_u8()? as usize;
                 let mut new_value = vec![0; len];
-                src.read_exact(&mut new_value[0..len]).await?;
+                buf.read_exact(&mut new_value[0..len])?;
 
-                let len = src.read_u8().await? as usize;
+                let len = buf.read_u8()? as usize;
                 let mut old_value = vec![0; len];
-                src.read_exact(&mut old_value[0..len]).await?;
+                buf.read_exact(&mut old_value[0..len])?;
 
                 TokenEnvChange::SqlCollation {
                     new: CollationInfo::new(new_value.as_slice()),
                     old: CollationInfo::new(old_value.as_slice()),
                 }
             }
             EnvChangeTy::BeginTransaction | EnvChangeTy::EnlistDTCTransaction => {
-                src.read_u8().await?;
-                let desc = src.read_u64_le().await?;
+                let len = buf.read_u8()?;
+                assert!(len == 8);
```

Comment:
:+1:

Reply:
I don't know if this is really needed. The JDBC code is extremely careful asserting all the things. The TDS standard dictates how the descriptor is always 8 bytes. And the only implementation is the SQL Server from Microsoft, which I highly doubt would send 9 bytes descriptor every now and then...